zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/ZiplineLoader.kt

```diff
@@ -136,114 +131,124 @@ class ZiplineLoader internal constructor(
-    val result = mutableListOf<Fetcher>()
-    val httpFetcher = HttpFetcher(httpClient, eventListener)
-    if (embeddedDir != null && embeddedFileSystem != null) {
-      result += FsEmbeddedFetcher(
-        embeddedDir = embeddedDir,
-        embeddedFileSystem = embeddedFileSystem,
-      )
-    }
-    if (cache != null) {
-      result += FsCachingFetcher(
-        cache = cache,
-        delegate = httpFetcher,
-      )
-    } else {
-      result += httpFetcher
+  suspend fun load(
+    applicationName: String,
+    manifestUrlFlow: Flow<String>,
+    oldestCodeToLoad: Duration = Duration.INFINITE,
+    initializer: (Zipline) -> Unit,
+  ): Flow<Zipline> {
+    return flow {
+      var first = true
+      var previousManifest: ZiplineManifest? = null
+
+      manifestUrlFlow.collect { manifestUrl ->
+        // Each time a manifest URL is emitted, download and initialize a Zipline for that URL.
+        //  - skip if the manifest hasn't changed from the previous load
+        //  - pin the application if the load succeeded; unpin if it failed.
+        withLifecycleEvents(applicationName, manifestUrl) {
+          val networkManifest = fetchManifestFromNetwork(applicationName, manifestUrl)
+          if (networkManifest.manifest == previousManifest) return@withLifecycleEvents // Unchanged.
+
+          try {
+            val networkZipline = loadFromManifest(applicationName, networkManifest, initializer)
+            cachingFetcher?.pin(applicationName, networkManifest) // Pin after successful loads.
+            emit(networkZipline)
+            previousManifest = networkManifest.manifest
+          } catch (e: Exception) {
+            cachingFetcher?.unpin(applicationName, networkManifest) // Unpin after failed loads.
+            throw e
+          }
+        }
+
+        // If network loading failed (due to network error, or bad code), attempt to load from the
+        // cache or embedded file system. This doesn't update pins!
+        if (previousManifest == null && first) {
+          first = false
+          val localManifest = loadCachedOrEmbeddedManifest(applicationName) ?: return@collect
+          withLifecycleEvents(applicationName, manifestUrl = null) {
+            val localZipline = loadFromManifest(applicationName, localManifest, initializer)
+            emit(localZipline)
+            previousManifest = localManifest.manifest
+          }
+        }
+      }
     }
-    result
   }
 
-  suspend fun loadOrFallBack(
+  suspend fun loadOnce(
     applicationName: String,
     manifestUrl: String,
+    oldestCodeToLoad: Duration = Duration.INFINITE,
     initializer: (Zipline) -> Unit = {},
   ): Zipline {
-    return try {
-      createZiplineAndLoad(applicationName, manifestUrl, null, initializer)
-    } catch (e: Exception) {
-      createZiplineAndLoad(applicationName, null, null, initializer)
-    }
+    return load(applicationName, flowOf(manifestUrl), oldestCodeToLoad, initializer)
+      .first()
```

Comment:
This throws the wrong exception if nothing loads: `NoSuchElementException`. Recommendations? Could do `firstOrNull()` and then throw an IllegalStateException. Or perhaps make this function return null?

Reply:
Fix coming!

Reply:
https://github.com/cashapp/zipline/pull/621